### PR TITLE
[IMPROVEMENT] Improve time prediction

### LINKED ISSUE #287: [PROPOSAL] Better Estimated Time Prediction
Please prefix your issue with one of the following: [BUG], [PROPOSAL], [QUESTION].

Sample platform commit (found at the bottom of each page) : **X.X.X**

**In raising this issue, I confirm the following (please check boxes, eg [X]):**

- [ ] I have read and understood the [contributors guide](https://github.com/CCExtractor/sample-platform/blob/master/.github/CONTRIBUTING.md).
- [ ] I have checked that the bug-fix I am reporting can be replicated, or that the feature I am suggesting isn't already present.
- [ ] I have checked that the issue I'm posting isn't already reported.
- [ ] I have checked that the issue I'm posting isn't already solved and no duplicates exist in [closed issues](https://github.com/CCExtractor/sample-platform/issues?q=is%3Aissue+is%3Aclosed) and in [opened issues](https://github.com/CCExtractor/sample-platform/issues)
- [ ] I have checked the pull requests tab for existing solutions/implementations to my issue/suggestion.

**My familiarity with the project is as follows (check one, eg [X]):**

- [ ] I have never visited/used the platform.
- [ ] I have used the platform just a couple of times.
- [ ] I have used the platform extensively, but have not contributed previously.
- [ ] I am an active contributor to the platform.

---

The current method (added via #a8d713) of test estimation uses the below formula to estimate the remaining time for the test.

Screenshot from 2019-06-03 23-11-18

An updated method as below will be used.

Screenshot from 2019-06-03 23-11-18

Below diagram represents the working of the new method.

Screenshot from 2019-06-03 23-17-52

The new method will requires the following:
- Store average preparation time for each platform alongside average time
- Add new fields in the sample table and store average time for each sample on
each platform
- While the test is running and reaches a particular sample, add a beginning
timestamp and similarly add an ending timestamp which will be used to calculate
it’s running time.
- The above newly added values are updated at the end of each test.

This should be done after #285

### PR DESCRIPTION
**Pull request by thealphadollar**
Please prefix your pull request with one of the following: **[FEATURE]** **[FIX]** **[IMPROVEMENT]**.

**In raising this pull request, I confirm the following (please check boxes):**

- [X] I have read and understood the [contributors guide](https://github.com/CCExtractor/sample-platform/blob/master/.github/CONTRIBUTING.md).
- [X] I have checked that another pull request for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**My familiarity with the project is as follows (check one):**

- [ ] I have never used the project.
- [ ] I have used the project briefly.
- [ ] I have used the project extensively, but have not contributed previously.
- [X] I am an active contributor to the project.

---

The Pull Request brings the following changes:

- Add database changes required for improving expected time
	- It is required to store preparation time and build time.
	- Migrations need to be run on the production database.
- Add logic for improving expected time
	- Add code for setting new KVM timings.
	- Save four new values in general data to keep in account build and prep time.
	- Update prediction feature to take into account the new data.

fixes #287 

NOTE: Tests will be added post review.

### CODE REVIEW

**Comment by thealphadollar** [comment]
Failing test will be fixed post review.

File: mod_ci/controllers.py [hunk 1]
```diff
@@ -1060,6 +1084,50 @@ def progress_reporter(test_id, token):
...
+
+
+def set_avg_build_time(platform: TestPlatform) -> None:
+    """
+    Set average platform preparation time.
+
+    :param platform: platform to which the average time belongs
+    :type platform: TestPlatform
+    """
+    pass
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: Is this unused, or still WIP?

File: mod_test/controllers.py [hunk 2]
```diff
@@ -99,6 +99,13 @@ def get_data_for_test(test, title=None) -> Dict[str, Any]:
     """
     if len(test.progress) == 0:
         var_average = 'average_time_' + test.platform.value
+
+        # get average build and prep time.
+        prep_average_key = test.platform.value + '_avg_prep_time'
```
Comment (canihavesomecoffee) [inline mod_test/controllers.py]: The ones above are `name_platform`, now you're switching around... Might be better to stick with one naming scheme :)

**Comment by thealphadollar** [comment]
@canihavesomecoffee Tests added.

**Pull request merged**